core/drivers/crypto/se050/session.c

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2020 - All Rights Reserved
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <config.h>
+#include <initcall.h>
+#include <se050.h>
+
+sss_se05x_key_store_t *se050_kstore;
+sss_se05x_session_t *se050_session;
+struct sss_se05x_ctx se050_ctx;
+
+TEE_Result se050_core_early_init(struct se050_scp_key *keys)
+{
+	sss_status_t status = kStatus_SSS_Success;
+
+	status = se050_session_open(&se050_ctx, keys);
+	if (status != kStatus_SSS_Success)
+		return TEE_ERROR_GENERIC;
+
+	if (IS_ENABLED(CFG_CORE_SE05X_INIT_NVM)) {
+		status = se050_factory_reset(&se050_ctx.session.s_ctx);
+		if (status != kStatus_SSS_Success)
+			return TEE_ERROR_GENERIC;
+	}
+
+	if (se050_ctx.session.subsystem == kType_SSS_SubSystem_NONE)
+		return TEE_ERROR_GENERIC;
+
+	status = se050_key_store_and_object_init(&se050_ctx);
+	if (status != kStatus_SSS_Success)
+		return TEE_ERROR_GENERIC;
+
+	se050_session = (sss_se05x_session_t *)((void *)&se050_ctx.session);
+	se050_kstore = (sss_se05x_key_store_t *)((void *)&se050_ctx.ks);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result display_info(void)
+{
+	se050_display_board_info(se050_session);
+	sss_se05x_session_close(se050_session);
+	return se050_core_early_init(NULL);
+}
+
+static TEE_Result enable_scp03(void)
+{
+	if (se050_enable_scp03(se050_session) != kStatus_SSS_Success)
+		return TEE_ERROR_GENERIC;
+
+	/*
+	 * Do not provision the keys at this point unless there is guaranteed
+	 * access to trusted storage.
+	 *
+	 * This can be done once RPMB is accessible and we can test it
+	 *
+	 * #if defined(CFG_CORE_SE05X_SCP03_PROVISION)
+	 *	if (se050_rotate_scp03_keys(&se050_ctx) != kStatus_SSS_Success)
+	 *		return TEE_ERROR_GENERIC;
+	 * #endif
+	 */
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result se050_early_init(void)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	ret = se050_core_early_init(NULL);
+
+	if (!ret && IS_ENABLED(CFG_CORE_SE05X_DISPLAY_INFO))
+		ret = display_info();
+
+	if (!ret && IS_ENABLED(CFG_CORE_SE05X_SCP03_EARLY))
+		return enable_scp03();
+
+	return ret;
```

Comment:
FYI: if reporting an error case here, Core will do nothing specific but printing an warning message.
If `ret != TEE_SUCCESS` is a major issue in your case, this should better panic the Core.
If not a major issue, then it's fine simply returning the error core.

Reply:
I still dont understand what  a major issue is.  ie, if the RNG stops working, is it worth a panic()? what about if the crypto device is bricked? I mean, optee can still do PM tasks for instance...so at what point do we pull the rug?

Reply:
Ok, fine. I undertand failing to initialize SE050 may not be a major issue. As far as OP-TEE can operate (with a reduced feature set) without later panicking then it's ok returning here with an error without panicking.

Reply:
I don't know what OP-TEE defines as being a major issue (I doubt OP-TEE has a clear definition either, do you?)....if not being able to open TAs is a major issue, then a crypto device implementing certain ciphers going down will certainly be a major issue. 
The way I see it, before we continue to have further discussions on when to panic, it would be nice to have some guideline...otherwise we are just working on opinions, are we not?

Reply:
I agree whether OP-TEE should panic or operate with a reduced feature set can product specific.
I don't have generic guidelines.
Here, if the platform embedding SE050 cannot operate without SE050 support, then it can be a good reason to panic on SE050 failing initialization. But it seems SE050 is not mandatory for OP-TEE core to operate.